### PR TITLE
feat(Modal): support keyboard navigation for scrolling content

### LINKED ISSUE #3261: AVT 2 (webkit) - React Transactional Modal & Input Modal -  Keyboard users can't scroll content
## Environment
macOS Mojave version 10.14.5
Version 75.0.3770.100 (Official Build) (64-bit)
Carbon v10 - React

## Detailed Description
Tab to the Launch Modal button 
Arrow keys do not scroll the content

### LINKED ISSUE #3816: [React Modal] carry over a11y updates from Vanilla
changes were made to modal for keyboard users that need carried over to react: 

https://github.com/carbon-design-system/carbon/pull/3650

### PR DESCRIPTION
**Pull request by emyarod**
Closes #3816
Closes #3261

This PR supports keyboard navigation for modal content when `hasScrollingContent` is specified. Refs #3650

#### Changelog

**New**

- `hasScrollingContent` prop for `<Modal>` and `<ComposedModal>`

#### Testing / Reviewing

Ensure keyboard navigation for modals and composed modals behaves as expected

### CODE REVIEW

**Comment by emyarod** [comment]
currently we require the user to specify if the modal contains scrolling content but would we want to determine this dynamically?

File: packages/react/src/components/ComposedModal/ComposedModal.js
Comment (abbeyhrt) [inline packages/react/src/components/ComposedModal/ComposedModal.js]: When the `tabIndex="0"`, I believe that `role="region"` and an aria label need to be added too

File: packages/react/src/components/Modal/Modal.js
Comment (abbeyhrt) [inline packages/react/src/components/Modal/Modal.js]: Same as above

File: packages/react/src/components/ComposedModal/ComposedModal.js
Comment (dakahn) [reply packages/react/src/components/ComposedModal/ComposedModal.js]: [Short note on improving usability of scrollable regions - Paciello Group](https://developer.paciellogroup.com/blog/2016/02/short-note-on-improving-usability-of-scrollable-regions/)

**Review by asudoh — APPROVED** [review]
LGTM 👍, given the code makes sequential-focusability optional. Thanks @emyarod!

**Review by joshblack — CHANGES_REQUESTED** [review]
Do we want this capability to be a prop, or a dynamic lookup on the node itself? e.g. checking `scrollHeight > clientHeight`

**Comment by asudoh** [comment]
> checking `scrollHeight > clientHeight`

This causes [synchronous layout query](https://developers.google.com/web/fundamentals/performance/rendering/avoid-large-complex-layouts-and-layout-thrashing#avoid_forced_synchronous_layouts) that has performance impact.

**Comment by joshblack** [comment]
@asudoh right, the caller would need to do something similar though, correct? I don't know if it's always statically known whether a region will be scrollable. The exact implementation doesn't need to be what is shared above (the synchronous check), but rather the idea that we could help people by detecting this and applying it on their behalf rather than them having to implement it.

Separately, I think we should careful saying querying for these properties result in layout thrashing by default. I believe the linked article suggests that if we access these properties and layout operations have been queued, we'll force layout by querying. Like it suggests in the article though, we can do style reads before any writes so that the browser can use the previous frame's layout values.

**Comment by asudoh** [comment]
If application author knows a modal has large content, they can use the prop this PR introduces. BTW I'm referring to synchronous layout query instead of layout thrashing.

**Comment by emyarod** [comment]
I updated the role and ARIA label settings but have we come to a consensus on whether or not to dynamically apply these?

**Comment by dakahn** [comment]
We have precedence set in previous work for this to be non-dynamic. I'm leaning toward that to get this merged. Maybe with an eye toward further work exploring how to make this dynamic _and_ performant, folks?

**Comment by asudoh** [comment]
@dakahn Makes sense to me!

**Review by dakahn — APPROVED** [review]

**Comment by joshblack** [comment]
@dakahn what I'm trying to suggest is that we should set folks up for success here versus having a burden of knowledge to know that this is required. Would you agree?

Separately, in the other thread I had asked if we should even fix this as Webkit has acknowledged that FF is the browser that is implementing this to spec and there seemed to be activity to address this in the associated Chromium trackers.

**Comment by elizabethsjudd** [comment]
@joshblack is there an ETA on the bugfix from Webkit?  Depending on the expected fix, it should be implemented in Carbon because it's an a11y issue and this ability to scroll is required for compliance.  Webkit browsers are a commonly used browser engine so it has a pretty big impact to have a patch until the higher-up fix can be applied.

**Comment by joshblack** [comment]
@elizabethsjudd I think what is most confusing is that the ticket is labelled as fixed and closed: https://bugs.chromium.org/p/chromium/issues/detail?id=585413

After digging through, it seems like some spec ambiguity is present which only makes this harder 😂 The exact diff in gerrit seems to be: https://chromium-review.googlesource.com/c/chromium/src/+/1258331 where the example seems to not use `tab-index` (although it's contained in the title) which also was confusing.

Going back to their intent to ship, there is also: https://groups.google.com/a/chromium.org/forum/#!msg/blink-dev/Fku6784p0wI/3MyOMTk7BwAJ which confirms that they wanted to change the behavior here such that we wouldn't have to do that.

Not able to dive too much more, figured @dakahn and the a11y project this cycle could investigate 👍

**Comment by scottnath** [comment]
Does Carbon or IBMa have a documented policy that would inform decisions like when to wait for a browser fix vs when it should be fixed in Carbon until browser support is there?

**Comment by joshblack** [comment]
@scottnath here's my quick attempt at defining it:

| Browser vendor                                                                      | Implement fix | Wait for upstream fix |
| :---------------------------------------------------------------------------------- | :-----------: | --------------------- |
| No bug reported to vendor, no intent to ship                                        |       X       |                       |
| Bug reported to vendor, no intent to ship                                           |       X       |                       |
| Bug reported to vendor, intent to ship, no progress on fix for more than 3 months, fix causes no adverse behavior |       X       |                       |
| Bug reported to vendor, intent to ship, no progress on fix for more than 3 months, fix causes adverse behavior    |               | X                     |
| Bug reported to vendor, intent to ship, active issue                                |               | X                     |

I think the tl;dr is that most of the time we'll want to implement a fix, only exceptions being that doing so would adversely affect other browsers or if the browser is actively addressing it.

**Comment by dakahn** [comment]
Okay then, we can put this PR in a holding pattern until I can look into this further. I'm currently working on another issue so give me until EOD Monday to figure out what's going on.

**Comment by joshblack** [comment]
@dakahn any updates?

**Comment by dakahn** [comment]
Blew past my self-imposed deadline. Sorry. With the peculiarities between the browsers and the _tough to tell_ state of this fix in Chromium I think we should just make this happen. That way we can reliable guarantee usability across browsers and OS's for our lib right now.

_As for the nature of the fix_, having this handled via prop and just changing our default example in Storybook or wherever to demonstrate proper use is fine by me. I know the proposed alternative (programmatically determining if text will overflow and making the text area a tab stop) is controversial for performance reasons -- do we have a good solution in mind to mitigate performance concerns? If we do then to @joshblack's point, yeah we should go with that.

**Comment by joshblack** [comment]
I think before we say it's a performance concern we should have criteria that non-trivial fixtures are used to verify and hopefully test against to prevent regressions in the future. Absent of data, it's hard to know if what we're doing is a premature optimization.

I care more about the experience here than anything else, how are we anticipating folks using a prop and determining when something should receive the prop or otherwise? How should it respond to changes in screen size?

**Comment by dakahn** [comment]
Default the prop to true? Worst case scenario there's an extra tab stop on the text content (but everything works).

**Comment by dakahn** [comment]
Maybe with a [:focus-visible polyfill](https://wicg.github.io/focus-visible/explainer.html)?

**Comment by joshblack** [comment]
Your call on defaulting to true, whatever makes the most sense 👍 Most likely should avoid that polyfill, unfortunately 😞

**Comment by emyarod** [comment]
@dakahn any ruling on the default value?

**Comment by dakahn** [comment]
Let's do it 👍

**Review by joshblack — CHANGES_REQUESTED** [review]
The storybook demo for Modal seems to have an invalid `aria-labelledby` when `hasScrollingContent` is enabled

image

I believe it needs to use an `id` in `aria-labelledby` instead of a selector.

**Comment by joshblack** [comment]
If this is on by default, would imply that `aria-label` or `aria-labelledby` would be required? @dakahn

**Comment by dakahn** [comment]
Yeah, looks like an `aria-label="scrolling text content"` or something like that works for NVDA on FF in Windows 10. Testing VO rn

**Comment by joshblack** [comment]
Any updates @dakahn? 👀

**Comment by dakahn** [comment]
@emyarod @joshblack Sure, yeah -- where are we on this? Defaulting to _on_ with the necessary aria tech?

**Comment by emyarod** [comment]
@dakahn I think that's what we were waiting for your confirmation on

**Comment by dakahn** [comment]
> Let's do it 👍

Let's do it 👍

**Comment by emyarod** [comment]
oh you wrote that you were still testing afterwards so I wasn't sure if your decision would still stand. in any case, instance IDs are now being generated and added to the modal label or heading and then applied as `aria-labelledby` on the scrollable content node which is scrollable by default now as well

**Comment by joshblack** [comment]
@laurenmrice for context this change will update the visual appearance of our modals such that the content area will receive a focus style

demo

File: packages/react/src/components/ComposedModal/ComposedModal.js
Comment (joshblack) [inline packages/react/src/components/ComposedModal/ComposedModal.js]: Would we consider this a breaking change since we are now requiring `aria-label` or `aria-labelledby` where-as we weren't before?

Comment (emyarod) [reply packages/react/src/components/ComposedModal/ComposedModal.js]: I guess we can consider it a breaking change, does that mean we want to put this behind a feature flag?

Comment (joshblack) [reply packages/react/src/components/ComposedModal/ComposedModal.js]: I'm definitely torn, I feel like we could get away with only supporting `aria-label` and then maybe doing a custom prop type like:

```ts
// Supply the prop type that, when available, means that
// this prop is required
type isRequiredWith = (propName: string, propType: PropType) => PropTypeValidator

// Usage
'aria-label': isRequiredWith('hasScrollingContent', PropTypes.string.isRequired)
```

What do you think? That way this prop is only required if `hasScrollingContent` is true. Unfortunately this might mean that this is off by default so not sure how @dakahn feels about that

Comment (emyarod) [reply packages/react/src/components/ComposedModal/ComposedModal.js]: I think I'd rather go with the custom validator instead of feature flag, similar to https://github.com/carbon-design-system/carbon/blob/master/packages/react/src/prop-types/requiredIfValueExists.js I guess?

**Review by laurenmrice — APPROVED** [review]
looks good to me 👍

File: packages/react/src/components/ComposedModal/ComposedModal.js
Comment (joshblack) [reply packages/react/src/components/ComposedModal/ComposedModal.js]: Yup! Would make sense

**Comment by joshblack** [comment]
@dakahn do you think you all could finish this up as part of a11y project this cycle? Is it still relevant?

**Comment by dakahn** [comment]
@joshblack Nah, we're not picking this up this cycle. 

Where is this PR? Can we merge this? What's the blocker?

**Comment by joshblack** [comment]
Remaining work is in: https://github.com/carbon-design-system/carbon/pull/3943#discussion_r336583172

I believe it was picked up last cycle by @abbeyhrt before realizing that this PR tackled it which is why I was curious if the a11y group wanted to try and tackle the last bit of work 👍

**Comment by emyarod** [comment]
is there still any unaddressed feedback since the last commit..?

**Review by abbeyhrt — APPROVED** [review]
Looks good to me!

**Review by joshblack — APPROVED** [review]

**Pull request merged**
